File: Animation.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace IsoBlockEditor
{
    public class Animation
    {
        Texture2D _spritesheet;
        Rectangle[] _frames;
        int _currentFrame;
        float _frameTime;
        bool _loop;

        public Animation(Texture2D spritesheet, Rectangle[] frames, bool loop = true)
        {
            _spritesheet = spritesheet;
            _frames = frames;
            _currentFrame = 0;
            _loop = loop;
        }

        public void Update(GameTime gameTime)
        {
            _frameTime += (float)gameTime.ElapsedGameTime.TotalSeconds;
            if (_frameTime > 0.1f)
            {
                _currentFrame++;
                if (_loop)
                {
                    if (_currentFrame >= _frames.Length) _currentFrame = 0;
                }
                else
                {
                    if (_currentFrame >= _frames.Length) _currentFrame = _frames.Length - 1;
                }
                _frameTime = 0;
            }
        }

        public void Draw(SpriteBatch spriteBatch, Rectangle destination, SpriteEffects effects)
        {
            spriteBatch.Draw(
                texture: _spritesheet,
                destinationRectangle: destination,
                sourceRectangle: _frames[_currentFrame],
                color: Color.White,
                rotation: 0,
                origin: Vector2.Zero,
                effects: effects,
                layerDepth: 0);
        }
    }
}

File: Animation.cpp
#include "Animation.h"



Animation::Animation(SDL_Texture* spritesheet, int startframe, int framecount, int framewidth, int frameheight, bool loop)
{
	Animation::spritesheet = spritesheet;
	Animation::frame = 0;
	Animation::startframe = startframe;
	Animation::framecount = framecount;
	Animation::loop = loop;
	Animation::frametime = 0.0f;
	Animation::framewidth = framewidth;
	Animation::frameheight = frameheight;

	create_frames();
}

Animation::~Animation()
{
	Animation::spritesheet = nullptr;
	for (int i = 0;i < framecount;i++)
	{
		delete (frames[i]);
	}
}

void Animation::create_frames()
{
	for (int i = 0;i < framecount;i++)
	{
		SDL_FRect* f = new SDL_FRect();
		f->w = (float)framewidth;
		f->h = (float)frameheight;
		f->x = (float)((startframe + i) * framewidth);
		f->y = 0.0f;
		frames.push_back(f);
	}
}

void Animation::update(double &deltatime)
{
	Animation::frametime += deltatime;
	if (frametime > 0.1f)
	{
		Animation::frame++;
		if (Animation::frame >= Animation::framecount)
		{
			frame = Animation::loop ? 0 : Animation::framecount - 1;
		}
		Animation::frametime = 0;
	}
}

void Animation::render(SDL_Renderer* renderer, SDL_FRect* destination)
{
	SDL_RenderTexture(renderer, spritesheet, frames[frame], destination);
}
